app/policies/wiki_policy.rb

```diff
@@ -6,7 +6,55 @@ def initialize(user, wiki)
     @wiki = wiki
   end
 
+  def show?
+      user.present? and (user.admin? or (user.premium? and user == wiki.user) or wiki.users.include?(user))
+  end
+
+  def edit?
+    update?
+  end
+
+  def update?
+    if wiki.private?
+      user.present? and (user.admin? or (user.premium? and user == wiki.user) or wiki.users.include?(user))
+    else
+      user.present?
+    end
+  end
+
   def destroy?
     user.present? and (user.admin? or user == wiki.user)
   end
+
+  class Scope
+    attr_reader :user, :scope
+
+    def initialize(user, scope)
+      @user = user
+      @scope = scope
+    end
+
+    def resolve
+      wikis = []
+      if user.present? and user.admin?
+        wikis = scope.all
+      elsif user.present? and user.premium?
+        all_wikis = scope.all
+        all_wikis.each do |wiki|
+          if wiki.private == false || user == wiki.user || wiki.users.include?(user)
+            wikis << wiki
+          end
+        end
+      else
+        all_wikis = scope.all
+        wikis = []
+        all_wikis.each do |wiki|
```

Comment:
There is lots of duplication here and is pretty complex. How could you refactor this to remove duplication and make this more readable?